### PR TITLE
Upload/Download Profile Data with CIP8

### LINKED ISSUE #6101: [P1] Profile Photos + Names
# What's important about this?
Users can see the profile photo and name of someone they are sending to
# What is "done"?
* Figma design for onboarding: https://www.figma.com/file/zt7aTQ5wuXycIwxq5oAmF9/Wallet-Refresh?node-id=2806%3A0
* Requirements: https://docs.google.com/document/d/15Vyai_T7r-6zeHdMx5DewC9oZ6AbPt66cVKoh2iEtcw/edit
* Backend design: https://docs.google.com/document/d/16VBKG68oTY5hUpjLUakYO788GfPEZyqKFSM15bu7Ca0/edit

### PR DESCRIPTION
**Pull request by isabellewei**
### Description

Upload and download profile picture+name using CIP8 in Valora. The user's own profile data is uploaded right after their DEK has been registered.

### Other changes

After the user's DEK has been registered, it is added to their geth native wallet, as the DEK is needed for encrypting and decrypting data.

### Tested

Tested using Android device and iOS emulator

### Related issues

- Fixes #6101 partially

### Backwards compatibility

Even if a user's DEK has already been registered previously, the app will check to see if their DEK has been added to the geth native wallet and if their profile info has been uploaded.

### CODE REVIEW

File: packages/mobile/src/account/profileInfo.ts [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+  async readDataFromAsResult<DataType>(
+    account: Address,
+    dataPath: string,
+    _checkOffchainSigners: boolean,
+    type?: t.Type<DataType>
+  ): Promise<Result<Buffer, OffchainErrors>> {
+    let dataResponse, signatureResponse
+
+    const accountRoot = `${config.CIP8MetadataUrl}/${toChecksumAddress(account)}`
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: This is an optimisation for Valora users, we should fallback to the usual offchain read function if it fails. 

We should also consider registering this metadata URL for Valora users, otherwise only other Valora users will know how to resolve ther name and profile picture. If they have already set their metadataURL we can consider aborting or overwriting it, I'm not sure which is the right solution.

File: packages/mobile/src/account/profileInfo.ts [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+// requires that the DEK has already been registered
+export function* uploadProfileInfo() {
+  const isAlreadyUploaded = yield select(isProfileUploadedSelector)
+  if (isAlreadyUploaded) {
+    return
+  }
+  try {
+    yield call(unlockDEK, true)
+    // yield call(addMetadataClaim)
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: Mentioned above but we probably do want to still do this.

File: packages/mobile/src/account/profileInfo.ts [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+// }
+
+export function* uploadNameAndPicture() {
+  const contractKit = yield call(getContractKit)
+  const account = yield select(currentAccountSelector)
+  const offchainWrapper = new UploadServiceDataWrapper(contractKit, account)
+
+  const name = yield select(nameSelector)
+  const nameAccessor = new PrivateNameAccessor(offchainWrapper)
+  let writeError = yield call([nameAccessor, 'write'], { name }, [])
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: Do we want to consider returning early if the nameAccessor write fails? Otherwise we lose the error if we also upload a profilePicture and that fails too.

File: packages/mobile/src/account/profileInfo.ts [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+  }
+
+  if (writeError) {
+    Logger.error(TAG + '@uploadNameAndPicture', writeError)
+    throw Error('Unable to write data')
+  }
+}
+
+// this function gives permission to the recipient to view the user's profile info
+export function* uploadSymmetricKeys(recipientAddresses: string[]) {
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: I'd also consider renaming this. References symmetric keys in the function name feels to low level for this API.

File: packages/mobile/src/account/profileInfo.ts [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+  // TODO: check if key for user already exists, skip if yes
+  const account = yield select(currentAccountSelector)
+  const contractKit = yield call(getContractKit)
+  yield call(unlockDEK)
+  const offchainWrapper = new UploadServiceDataWrapper(contractKit, account)
+
+  const nameAccessor = new PrivateNameAccessor(offchainWrapper)
+  let writeError = yield call([nameAccessor, 'allowAccess'], recipientAddresses)
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: Same issue here. Generally I prefer to return as early as possible from functions. With this the log line on 266 will still fire even though there could have been two errors.

Maybe something like this works a bit more explicitly

```
const nameAccessor = new PrivateNameAccessor(offchainWrapper)
const nameWriteError = yield call([nameAccessor, 'allowAccess'], recipientAddresses)
if (nameWriteError) {
  // log
  return // or throw
}

const pictureUri = yield select(pictureSelector)
if (pictureUri) {
  const pictureAccessor = new PrivatePictureAccessor(offchainWrapper)
  const pictureWriteError = yield call([pictureAccessor, 'allowAccess'], recipientAddresses)
  if (pictureWriteError) {
    // log
    return // or throw
  }
}

Logger.info(TAG + '@uploadSymmetricKeys', 'uploaded symmetric keys for ' + recipientAddresses)
```

File: packages/mobile/src/account/profileInfo.ts [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    throw new Error('No data key in store. Should never happen.')
+  }
+  const dataKeyaddress = normalizeAddressWith0x(
+    privateKeyToAddress(ensureLeading0x(privateDataKey))
+  )
+  const wallet: UnlockableWallet = yield call(getWallet)
+  if (addAccount) {
+    try {
+      yield call([wallet, wallet.addAccount], privateDataKey, '')
+    } catch (e) {
```
Comment (AlexBHarley) [inline packages/mobile/src/account/profileInfo.ts]: I'd imagine we want to log this error just in case something unexpected happens

File: packages/mobile/src/geth/GethNativeBridgeSigner.ts [hunk 7]
```diff
@@ -102,12 +103,20 @@ export class GethNativeBridgeSigner implements Signer {
     return this.unlockTime + this.unlockDuration - this.unlockBufferSeconds > currentTimeInSeconds()
   }
 
-  decrypt(_ciphertext: Buffer) {
-    return Promise.reject(new Error('Decryption operation is not supported on this signer'))
+  decrypt(ciphertext: Buffer): Promise<Buffer> {
+    const ciphertextBase64 = ciphertext.toString('base64')
+    return this.geth.decrypt(this.account, ciphertextBase64).then((textBase64) => {
```
Comment (AlexBHarley) [inline packages/mobile/src/geth/GethNativeBridgeSigner.ts]: Not 100% knowledgeable of our coding styles but I personally,

1. Prefer async await
2. Wouldn't use a return statement + curly brackets for a one liner function body

```
async decrypt(ciphertext: Buffer): Promise<Buffer> {
  const textBase64 = await this.geth.decrypt(this.account, ciphertext.toString('base64')
  return Buffer.from(textBase64, 'base64')
}
```

Seems a little cleaner to me.

File: packages/mobile/src/utils/image.ts [hunk 8]
```diff
@@ -9,6 +9,14 @@ const mimeTypeToExtension: { [key: string]: string | undefined } = {
   'image/jpg': 'jpg', // image/jpg is technically not a mime type, but handling it just in case.
 }
 
+export const extensionToMimeType: { [key: string]: string | undefined } = {
```
Comment (AlexBHarley) [inline packages/mobile/src/utils/image.ts]: Could consider just using https://www.npmjs.com/package/mime-types.

**Review by AlexBHarley — CHANGES_REQUESTED** [review]
Nice! Really excited for this to land.

Generally just some open questions on what we're doing to ensure other players are supported when we roll this out.

File: packages/mobile/ios/Podfile.lock [hunk 9]
```diff
@@ -1,7 +1,7 @@
 PODS:
   - Analytics (4.1.2)
   - boost-for-react-native (1.63.0)
-  - CeloBlockchain (0.0.319)
+  - CeloBlockchain (0.0.322)
```
Comment (jeanregisser) [inline packages/mobile/ios/Podfile.lock]: We need a published version that includes the [mobile patch](https://github.com/celo-org/celo-blockchain/commit/4d75d29209bd87e95729fb85e766d65cdcd2fd24) for the crash happening on Android x86_64.

This currently means `>= 0.0.336` (https://github.com/celo-org/celo-blockchain/commit/d14b756). We should check with the blockchain team what they recommend too.

cc @oneeman @gastonponti

File: packages/mobile/ios/Podfile.lock [hunk 10]
```diff
@@ -913,7 +913,7 @@ SPEC CHECKSUMS:
   Flipper-RSocket: 64e7431a55835eb953b0bf984ef3b90ae9fdddd7
   FlipperKit: 0367d53d5c2829ea8137eaa5d1296c5fc623848b
   Folly: b73c3869541e86821df3c387eb0af5f65addfab4
-  glog: 40a13f7840415b9a77023fbcae0f1e6f43192af3
+  glog: 1f3da668190260b06b429bb211bfbee5cd790c28
```
Comment (jeanregisser) [inline packages/mobile/ios/Podfile.lock]: CircleCI complains about this.
This means these podspec are different between CircleCI and your dev machine.
`bundle exec pod install --repo-update` should help. But @tarikbellamine reported it didn't fix the issue for him.
If this doesn't work, clearing the podspec cache could be tried too.
https://gist.github.com/mbinna/4202236

File: packages/mobile/src/account/profileInfo.ts [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+//     ])
+//     const setAccountTx = accountsWrapper.setMetadataURL(`${BUCKET_URL}${account}/metadata.json`)
+//     const context = newTransactionContext(TAG, 'Set metadata URL')
+//     yield call(sendTransaction, setAccountTx.txo, account, context)
+//     Logger.info(TAG + '@addMetadataClaim' + 'set metadata URL')
+//   } catch (error) {
+//     Logger.error(`${TAG}/addMetadataClaim`, 'Could not add metadata claim', error)
+//     throw error
+//   }
+// }
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: When should we do this?

File: packages/mobile/src/account/profileInfo.ts [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+import { isProfileUploadedSelector, nameSelector, pictureSelector } from 'src/account/selectors'
+import config from 'src/geth/networkConfig'
+import { extensionToMimeType, getDataURL, saveRecipientPicture } from 'src/utils/image'
+import Logger from 'src/utils/Logger'
+import { getContractKit, getWallet } from 'src/web3/contracts'
+import { currentAccountSelector, dataEncryptionKeySelector } from 'src/web3/selectors'
+
+const TAG = 'account/profileInfo'
+
+async function makeCall(data: any, signature: string): Promise<SignedPostPolicyV4Output[]> {
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: `makeCall` seems a bit too generic. Maybe `authorizeUrls`?

File: packages/mobile/src/account/profileInfo.ts [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+          method: 'POST',
+          headers: formData.getHeaders(),
+          // Use getBuffer() which ends up transferring the body as base64 data over the RN bridge
+          // because RN doesn't support Buffer inside FormData
+          // See https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/convertRequestBody.js#L34
+          // and https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/FormData.js
+          body: formData.getBuffer(),
+        }).then((x) => {
+          if (!x.ok) {
+            Logger.error(TAG + '@writeDataTo', 'Error uploading ' + x.headers.get('location'))
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: Should we throw here?

File: packages/mobile/src/account/profileInfo.ts [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+          }
+          return x.text()
+        })
+      })
+    )
+  }
+
+  async readDataFromAsResult<DataType>(
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: `readDataFromResult` or `readDataAsResult`?

File: packages/mobile/src/account/profileInfo.ts (hunk 1 above)
Comment (isabellewei) [reply packages/mobile/src/account/profileInfo.ts]: do you think this should be implemented in this PR? I was thinking it would be added in later

Comment (AlexBHarley) [reply packages/mobile/src/account/profileInfo.ts]: I think we can get away without it in this PR yeah. Should be top priority though once this is shipped and stable.

cc @aslawson

Comment (aslawson) [reply packages/mobile/src/account/profileInfo.ts]: EDIT: Note -- posted this late for documentation reasons. Read second comment below for context.
RE: Valora users registering the metadata onchain. 
I want to make sure the nomenclature isn't throwing me off -- so is 'CIP8MetadataUrl' the actual metadataURL which is wrapping the cip8 storage claim url? Or is this the direct url to the cip8 storage root?
This feels like it would be a bit tricky to backfill for users that don't register onchain initially. And if this is indeed the metadataURL, feels like a light lift to add Account.setMetadataURL() and we could then get readable profile+names across applications.

The considerations would be 
1. Edge case @AlexBHarley mentioned if they already have a metadata url registered. I think this would be extremely rare since only validators use this for attestation service registration and would likely not use that account for Valora (it's definitely inadvisable for other reasons). Otherwise, since name+profile pictures is the only existing use case, overwriting would seem fine. We'd just need to keep that in mind as we expand the use cases and capabilities of cip8. but should leave that call to the engineers ;)
2. long term, we would likely want to storage roots to be stored in contracts directly to support different sources of storage data (not just clab's hosted). seems we wouldn't be able to elegantly backfill for this storage anyway...
@nambrot to add any other thoughts.

overall, if the lift is indeed light, would be a nice add to have available read access. and feels we are still several steps away from supporting easy storage write access without overwriting, but seems ok as a short term solution.

File: packages/mobile/src/account/profileInfo.ts (hunk 6 above)
Comment (jeanregisser) [reply packages/mobile/src/account/profileInfo.ts]: Yes we should log the real error.

File: packages/mobile/src/geth/GethNativeBridgeSigner.ts (hunk 7 above)
Comment (jeanregisser) [reply packages/mobile/src/geth/GethNativeBridgeSigner.ts]: +1 to @AlexBHarley comment.

File: packages/sdk/contractkit/package.json [hunk 15]
```diff
@@ -53,5 +53,8 @@
   },
   "engines": {
     "node": ">=8.13.0"
-  }
+  },
+  "browser":{
+    "child_process": false
+  }
```
Comment (jeanregisser) [inline packages/sdk/contractkit/package.json:59]: Is this change intended?

File: packages/mobile/src/utils/image.ts [hunk 16]
```diff
@@ -32,3 +40,7 @@ export const saveProfilePicture = async (dataUrl: string): Promise<string> => {
     `file://${RNFS.DocumentDirectoryPath}/profile-${Date.now()}`
   )
 }
+
+export const saveRecipientPicture = async (dataUrl: string, address: string): Promise<string> => {
+  return saveImageDataUrlToFile(dataUrl, `file://${RNFS.DocumentDirectoryPath}/${address}`)
```
Comment (jeanregisser) [inline packages/mobile/src/utils/image.ts]: Looks like we should store them in a dedicated folder.
This way we can more easily clean up pictures if needed, or list them.

**Review by jeanregisser — COMMENTED** [review]
Awesome!! 🚀 Looking great overall! 
Could you add/update tests for the added/updated functions? In particular around `profileInfo.ts` and the changed sagas.
I realize `web3/dataEncryptionKey.ts` has currently no tests 😢 so adding tests for that one might be non trivial. In that case we should open an issue and come back to it later.

File: packages/mobile/src/account/profileInfo.ts (hunk 1 above)
Comment (aslawson) [reply packages/mobile/src/account/profileInfo.ts]: Realized I didn't publish this ^ yesterday, so just published to document original thoughts.

Talked with @AlexBHarley and @nambrot and it sounds like current state is that this is calling the storage url directly, so there is additional work still to create the metadata blobs.

I think it's ok as fast follow if Valora provides a way for users to later "share to wider network", so there is a route for them those to register on-chain if they add their photos/names before we have that implemented.

But would certainly like to prioritize it :)
https://github.com/celo-org/celo-monorepo/issues/6941

File: packages/mobile/src/account/profileInfo.ts (hunk 14 above)
Comment (AlexBHarley) [reply packages/mobile/src/account/profileInfo.ts]: @jeanregisser this is due to naming in the SDK Identity package. Kind of a weird name I agree.

`readDataFrom(path: string)` makes sense, the `AsResult` is a more developer friendly API where we don't throw and instead return an `Either<Error, Result>` type object.

File: packages/mobile/src/utils/image.ts (hunk 16 above)
Comment (AlexBHarley) [reply packages/mobile/src/utils/image.ts]: Not sure what you're suggesting here but our structure needs to be compatible with CIP8.

`<BUCKET>/<WHATEVER_PATH>/<USER_ADDRESS>/account/picture`.

File: packages/mobile/src/account/profileInfo.ts (hunk 11 above)
Comment (isabellewei) [reply packages/mobile/src/account/profileInfo.ts]: There's an issue tracking this: #6941

File: packages/mobile/src/account/profileInfo.ts (hunk 13 above)
Comment (isabellewei) [reply packages/mobile/src/account/profileInfo.ts]: it would be captured in `uploadNameAndPicture` or `giveProfileAccess` because of the `writeError`. Do you think we should throw directly in here too?

File: packages/sdk/contractkit/package.json (hunk 15 above)
Comment (isabellewei) [reply packages/sdk/contractkit/package.json:59]: yes! I forgot where exactly, but there's a line in contractkit that doesn't work in the app if this isn't included

File: packages/mobile/src/utils/image.ts (hunk 16 above)
Comment (isabellewei) [reply packages/mobile/src/utils/image.ts]: @AlexBHarley this is the path for where the pictures are stored internally in Valora after fetching them from the CIP8 backend

File: packages/mobile/src/account/profileInfo.ts (hunk 13 above)
Comment (jeanregisser) [reply packages/mobile/src/account/profileInfo.ts]: Ok, I'm not following entirely. What I see is that we just log the error but we still return `x.text()` 
Which indicates we're not returning the actual error. But maybe I'm missing something.

File: packages/mobile/src/account/profileInfo.ts [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+        path: `${dataPath}.signature`,
+      },
+    ]
+
+    const hexPayload = ensureLeading0x(
+      Buffer.from(JSON.stringify(signedUrlsPayload)).toString('hex')
+    )
+    const authorization = await this.kit.getWallet().signPersonalMessage(this.signer, hexPayload)
+    const signedUrls = await makeCall(signedUrlsPayload, authorization)
+    await Promise.all(
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: It seems we should wrap this like in https://github.com/celo-org/celo-monorepo/blob/6148646a29e5e719df9a1b32d1580ea7e6e61be8/packages/sdk/identity/src/offchain-data-wrapper.ts#L104-L111

File: packages/mobile/src/account/profileInfo.ts [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+  if (!privateDataKey) {
+    throw new Error('No data key in store. Should never happen.')
+  }
+  const dataKeyaddress = normalizeAddressWith0x(
+    privateKeyToAddress(ensureLeading0x(privateDataKey))
+  )
+  const wallet: UnlockableWallet = yield call(getWallet)
+  if (addAccount) {
+    try {
+      yield call([wallet, wallet.addAccount], privateDataKey, '')
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]: Though currently the data encryption key was stored in an unprotected way in the redux state.
We now have the possibility to secure it here and should not use an empty passphrase.
It seems we could protect it the same way we protect the main account key (PIN + generated password).
Thoughts @nategraf @nambrot @AlexBHarley @cmcewen?

File: packages/mobile/src/account/profileInfo.ts [hunk 19]
```diff
@@ -0,0 +1,323 @@
...
+    Logger.warn("can't fetch name for", address)
+  }
+}
+
+function* unlockDEK(addAccount = false) {
+  const privateDataKey: string | null = yield select(dataEncryptionKeySelector)
+  if (!privateDataKey) {
+    throw new Error('No data key in store. Should never happen.')
+  }
+  const dataKeyaddress = normalizeAddressWith0x(
```
Comment (jeanregisser) [inline packages/mobile/src/account/profileInfo.ts]:
```suggestion
  const dataKeyAddress = normalizeAddressWith0x(
```

File: packages/mobile/src/account/profileInfo.ts (hunk 18 above)
Comment (nambrot) [reply packages/mobile/src/account/profileInfo.ts]: Absolutely agree that Valora should not be storing the keys unprotected in redux

Comment (nategraf) [reply packages/mobile/src/account/profileInfo.ts]: I'd agree we should add some protection if possible. In order to support good UX, we should make sure it's still accessible without the PIN though because it's used so often. (We also would not keep the PIN in memory if possible, but that one is a bit more debatable given how low-entropy PINs are anyway....)

Comment (isabellewei) [reply packages/mobile/src/account/profileInfo.ts]: I also don't think we should require a PIN for the DEK, it would be really confusing for the user to remember two PINs and when to input which one. Is it possible to add a password to the DEK without a PIN?

Comment (isabellewei) [reply packages/mobile/src/account/profileInfo.ts]: Issue for removing DEK from redux: https://github.com/celo-org/wallet/issues/17

File: packages/sdk/utils/src/ecies.ts [hunk 20]
```diff
@@ -144,7 +144,7 @@ export function Encrypt(pubKeyTo: Buffer, plaintext: Buffer) {
   const px = ephemPrivKey.derive(
     ec.keyFromPublic(Buffer.concat([Buffer.from([0x04]), pubKeyTo])).getPublic()
   )
-  const hash = ConcatKDF(px.toBuffer(), 32)
+  const hash = ConcatKDF(px.toArrayLike(Buffer), 32)
```
Comment (medhakothari) [inline packages/sdk/utils/src/ecies.ts:147]: Could you explain the error that this line was causing before, curious

File: packages/sdk/identity/src/offchain/accessors/binary.ts [hunk 21]
```diff
@@ -45,6 +45,10 @@ export class PrivateBinaryAccessor implements PrivateAccessor<Buffer> {
     return writeEncrypted(this.wrapper, this.dataPath, data, toAddresses, symmetricKey)
   }
 
+  allowAccess(toAddresses: Address[], symmetricKey?: Buffer) {
```
Comment (medhakothari) [inline packages/sdk/identity/src/offchain/accessors/binary.ts]: how come an `async` isn't needed here?

File: packages/sdk/identity/src/offchain/accessors/simple.ts [hunk 22]
```diff
@@ -80,6 +86,10 @@ export class PrivateSimpleAccessor<DataType> implements PrivateAccessor<DataType
     return writeEncrypted(this.wrapper, this.dataPath, serialize(data), toAddresses, symmetricKey)
   }
 
+  allowAccess(toAddresses: Address[], symmetricKey?: Buffer) {
```
Comment (medhakothari) [inline packages/sdk/identity/src/offchain/accessors/simple.ts]: same as above

File: packages/sdk/identity/src/offchain/accessors/binary.ts (hunk 21 above)
Comment (medhakothari) [reply packages/sdk/identity/src/offchain/accessors/binary.ts]: Also, could you add a test in `offchain-data-wrapper.test.ts` that uses this function?

**Review by medhakothari — COMMENTED** [review]
Generally looks good to me, but added a few questions/comments because I'm not super familiar w CIP 8

File: packages/sdk/utils/src/ecies.ts (hunk 20 above)
Comment (isabellewei) [reply packages/sdk/utils/src/ecies.ts:147]: I did this such a long time ago that I could not remember haha, but after rooting around in my slack message history, it seems that using `toBuffer()` in a react app results in a `Assertion failed` error

File: packages/sdk/identity/src/offchain/accessors/binary.ts (hunk 21 above)
Comment (isabellewei) [reply packages/sdk/identity/src/offchain/accessors/binary.ts]: hm seems like there should be a `async`, but it seemed to be working without. I've added it anyways

File: packages/sdk/utils/src/ecies.ts (hunk 20 above)
Comment (medhakothari) [reply packages/sdk/utils/src/ecies.ts:147]: ooh interesting, thx!

**Review by medhakothari — APPROVED** [review]
LGTM! thx for adding the changes

**Review by AlexBHarley — APPROVED** [review]
Very nice!
